File: js/test_money.js
const assert = require('assert')
const Money = require('./money')
const Portfolio = require('./portfolio')
// Chapter 1. Create Dollar class instance
// Chapter 2. Replace Dollar with money to keep it DRY

const fiveDollar = new Money(5)
const tenDollar = fiveDollar.times(2)
const fifteenDollar = fiveDollar.times(3)

assert.strictEqual(tenDollar.amount, 10, '5 Dollar * 2 should equal 10 Dollar')
console.log('testing 5 Dollar * 2 should equal 10 Dollar passed')
assert.strictEqual(fifteenDollar.amount, 15, '5 Dollar * 3 should equal 15 Dollar')
console.log('testing 5 Dollar * 3 should equal 15 Dollar passed')

// Chapter 2. Create Money Class instance for different currency
let tenEuros = new Money(10, 'EUR')
let twentyEuros = tenEuros.times(2)
assert.strictEqual(twentyEuros.amount, 20)
console.log('testing 10 EUR * 2 should equal 20 EUR passed')
assert.strictEqual(twentyEuros.currency, 'EUR')
console.log('testing currency should be EUR passed')

// Chapter 2. Divide and Conquer
let originalMoney = new Money(4002, 'KRW')
let actualMoneyAfterDivision = originalMoney.divide(4)
let expectedMoneyAfterDivision = new Money(1000.5, 'KRW')
assert.deepStrictEqual(actualMoneyAfterDivision, expectedMoneyAfterDivision)
console.log('testing divide function passed')

// Chapter 3. 相加兩個不同匯率的 Money 物件，因此新增 instance 來處理此問題
let fifteenDollars = new Money(15, 'USD')
let portfolio = new Portfolio()
portfolio.add(fiveDollar, tenDollar)
assert.deepStrictEqual(portfolio.evaluate('USD'), fifteenDollars)
console.log('testing adding two dollar passed')

File: py/test_money.py
import unittest
# Step 03. Separating Our Code into Modules
from money import Money
from portfolio import Portfolio

class TestMoney(unittest.TestCase): 
        
    # def testMultiplicationInEuros(self): 
    #     tenEuros = Money(10, "EUR") 
    #     twentyEuros = tenEuros.times(2) 
    #     self.assertEqual(20, twentyEuros.amount)
    #     self.assertEqual("EUR", twentyEuros.currency)
        
    # def testMultiplicationInDollars(self):
    #     fiveDollars = Money(5, "USD")
    #     tenDollars = Money(10, "USD")
    #     self.assertEqual(tenDollars, fiveDollars.times(2))
    
    # # Step 05. Removing Redundancy in Tests
    def testMultiplication(self): 
        tenEuros = Money(10, "EUR") 
        twentyEuros = tenEuros.times(2) 
        self.assertEqual(20, twentyEuros.amount)
        self.assertEqual("EUR", twentyEuros.currency)
        
    def testDivision(self): 
        originalMoney = Money(4002, "KRW") 
        actualMoneyAfterDivision = originalMoney.divide(4)
        expectedMoneyAfterDivision = Money(1000.5, "KRW")
        
        self.assertEqual(expectedMoneyAfterDivision.amount, actualMoneyAfterDivision.amount)
        self.assertEqual(expectedMoneyAfterDivision.currency, actualMoneyAfterDivision.currency)
    
    def testAddition(self):
        fiveDollars = Money(5, "USD") 
        tenDollars = Money(10, "USD") 
        fifteenDollars = Money(15, "USD") 
        portfolio = Portfolio() 
        portfolio.add(fiveDollars, tenDollars)
        self.assertEqual(fifteenDollars, portfolio.evaluate("USD"))



if __name__ == '__main__': 
    unittest.main()